src/happy.rs

```diff
@@ -0,0 +1,95 @@
+use std::io;
+use std::iter::FusedIterator;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::mpsc::{channel, RecvTimeoutError};
+use std::thread;
+use std::time::{Duration, Instant};
+
+const DEFAULT_CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+const RACE_DELAY: Duration = Duration::from_millis(200);
+
+/// This function implements a basic form of the happy eyeballs RFC to quickly connect
+/// to a domain which is available in both IPv4 and IPv6. Connection attempts are raced
+/// against each other and the first to connect successfully wins the race.
+///
+/// If the timeout is not provided, a default timeout of 10 seconds is used.
+pub fn connect<A>(addrs: A, timeout: impl Into<Option<Duration>>) -> io::Result<TcpStream>
+where
+    A: ToSocketAddrs,
+{
+    let timeout = timeout.into().unwrap_or(DEFAULT_CONNECTION_TIMEOUT);
+    let addrs: Vec<_> = addrs.to_socket_addrs()?.collect();
+    let ipv4 = addrs.iter().filter(|a| a.is_ipv4()).cloned();
+    let ipv6 = addrs.iter().filter(|a| a.is_ipv6()).cloned();
+    let order = intertwine(ipv4, ipv6);
+
+    let (tx, rx) = channel();
+    let mut last_err = None;
+
+    let start = Instant::now();
+
+    for addr in order {
+        let tx = tx.clone();
+
+        thread::spawn(move || {
+            debug!("trying to connect to {}", addr);
+
+            let _ = tx.send(TcpStream::connect_timeout(&addr, timeout));
+        });
+
+        match rx.recv_timeout(RACE_DELAY) {
+            Ok(Ok(sock)) => {
+                let end = Instant::now();
+                let span = end - start;
+                debug!("success, took {}ms", span.as_millis());
+
+                return Ok(sock);
+            }
+            Ok(Err(err)) => {
+                debug!("connection error: {} addr={}", err, addr);
+
+                last_err = Some(err);
+            }
+            Err(RecvTimeoutError::Timeout) => {
+                continue;
```

Comment:
Don't we still need a fully blocking `recv` after the loop? E.g. we spawn two threads but nothing connects in 400ms and hence we bail out, but the first thread could just succeed after 500ms? So I think a second call to `recv` is necessary.

Reply:
Yeah you're right. I think I have to loop after the for loop in case they all timed out.

Reply:
I added a second loop to ensure we catch everything, let me know what you think.

Reply:
I think you could avoid `deadline` und just use `recv` instead of `recv_timeout` as the threads will just hang up until they timed out you will break the loop due to the `Err(_)` match arm.

Reply:
Yeah you're absolutely right. At first I forgot to drop the tx so the disconnection was not happening. But I added the drop later and now it does occur so the timeout is not necessary.